core/src/main/java/com/linecorp/armeria/server/docs/DocService.java

```diff
@@ -259,13 +237,130 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
+                    ServiceSpecification spec = generate(services);
+                    spec = docStringSupport.addDocStrings(spec);
+                    spec = exampleSupport.addExamples(spec);
+                    try {
+                        final byte[] content = jsonMapper.writerWithDefaultPrettyPrinter()
+                                                         .writeValueAsBytes(spec);
+                        return toFile(content, MediaType.JSON_UTF_8);
+                    } catch (JsonProcessingException e) {
+                        throw new RuntimeException(e);
+                    }
+                }, executor);
+            });
+        }
+
+        void updateServices(List<ServiceConfig> services) {
+            servicesRef.set(services);
+        }
+
+        private static AggregatedHttpFile toFile(byte[] content, MediaType mediaType) {
+            return AggregatedHttpFile
+                    .builder(HttpData.wrap(content))
+                    .contentType(mediaType)
+                    .cacheControl(ServerCacheControl.REVALIDATED)
+                    .build();
+        }
+
+        private ServiceSpecification generate(List<ServiceConfig> services) {
+            return ServiceSpecification.merge(
+                    plugins.stream()
+                           .map(plugin -> plugin.generateSpecification(
+                                   findSupportedServices(plugin, services), filter, namedTypeInfoProvider))
+                           .collect(toImmutableList()));
+        }
+
+        private static NamedTypeInfoProvider composeNamedTypeInfoProvider(
+                @Nullable NamedTypeInfoProvider namedTypeInfoProvider) {
+            return typeDescriptor -> {
+                if (namedTypeInfoProvider != null) {
+                    // Respect user-defined provider first.
+                    final NamedTypeInfo namedTypeInfo = namedTypeInfoProvider.newNamedTypeInfo(typeDescriptor);
+                    if (namedTypeInfo != null) {
+                        return namedTypeInfo;
+                    }
+                }
+
+                for (NamedTypeInfoProvider provider : spiNamedTypeInfoProviders) {
+                    final NamedTypeInfo namedTypeInfo = provider.newNamedTypeInfo(typeDescriptor);
+                    if (namedTypeInfo != null) {
+                        return namedTypeInfo;
+                    }
+                }
+                return null;
+            };
+        }
+    }
+
     static final class DocServiceVfs extends AbstractHttpVfs {
+        private final SpecificationLoader specificationLoader;
+
+        DocServiceVfs(SpecificationLoader specificationLoader) {
+            this.specificationLoader = specificationLoader;
+        }
 
         private final HttpVfs staticFiles = HttpVfs.of(DocService.class.getClassLoader(),
                                                        "com/linecorp/armeria/server/docs");
```

Comment:
nit: Move the constructor below `staticFiles`?

Reply:
Thanks, updated!